ART-9183 Add rclone() wrapper function

locriandev opened the pull request:
rclone is using a default configuration file stored at /home/jenkins/.config/rclone/rclone.conf. This file currently carries AWS secrets in clear text. This PR proposes a different approach:
- store the secrets in Jenkins (already done)
- turn the config file into a template, where the secrets are replaced with env vars
- at runtime, render the template by filling the placeholders with real values coming from Jenkins (use `envsubst` for this`
- call `rclone` with the `--config` option, pointing to the locally rendered file
- when done, remove the rendered file (do that in a `finally` block to make sure this always happens)

joepvd commented on jobs/build/set_client_latest/S3-set-v4-client-latest.sh:
```diff
@@ -58,14 +67,18 @@ function transferClientIfNeeded() {
         # detects either, just copy (don't sync as it will try to be smart and not copy files with the same name & size).
-        rclone copy "${S3_SRC}" "${S3_DEST}"  # Copy over all files, regardless of the difference detected
-        rclone sync -c "${S3_SRC}" "${S3_DEST}"  # Run sync to delete any files that should no longer be present.
+        rclone(copy "${S3_SRC}" "${S3_DEST}")  # Copy over all files, regardless of the difference detected
```
```suggestion
        rclone copy "${S3_SRC}" "${S3_DEST}"  # Copy over all files, regardless of the difference detected
```

joepvd commented on jobs/build/set_client_latest/S3-set-v4-client-latest.sh:
```diff
@@ -58,14 +67,18 @@ function transferClientIfNeeded() {
     CHECK_RESULT="$?"
     # If non-multipart files in the directories are out of sync, then check will return an error
-    if ! rclone check "${S3_SRC}" "${S3_DEST}" || [[ "${FORCE_UPDATE}" == "1" ]]; then
+    if ! rclone(check "${S3_SRC}" "${S3_DEST}" || [[ "${FORCE_UPDATE}" == "1" ]]); then
```
```suggestion
    if ! rclone check "${S3_SRC}" "${S3_DEST}" || [[ "${FORCE_UPDATE}" == "1" ]]; then
```

thegreyd commented:
Awesome work! Can we get a job run in action with this script, I think --dry-run should work?

locriandev commented:
> Awesome work! Can we get a job run in action with this script, I think --dry-run should work?

@thegreyd yeah I've added a test Jenkinsfile (on a different branch, not to pollute this PR) that runs an `rclone ls`. I have moved the legacy rclone config file to a backup location, and ran https://art-jenkins.apps.prod-stable-spoke1-dc-iad2.itup.redhat.com/job/hack/job/dpaolell/job/test-rclone/10/console which succeeded.

thegreyd commented:
awesome
/lgtm

joepvd commented on jobs/build/set_client_latest/S3-set-v4-client-latest.sh:
```diff
@@ -62,10 +71,14 @@ function transferClientIfNeeded() {
         # detects either, just copy (don't sync as it will try to be smart and not copy files with the same name & size).
         rclone copy "${S3_SRC}" "${S3_DEST}"  # Copy over all files, regardless of the difference detected
-        rclone sync -c "${S3_SRC}" "${S3_DEST}"  # Run sync to delete any files that should no longer be present.
+        rclone(sync -c "${S3_SRC}" "${S3_DEST}")  # Run sync to delete any files that should no longer be present.
```
```suggestion
        rclone sync -c "${S3_SRC}" "${S3_DEST}"  # Run sync to delete any files that should no longer be present.
```

joepvd reviewed (approved):
/lgtm
